### PR TITLE
Fix #1286 (this time in the model)

### LINKED ISSUE #1286: If pledge sent, pledge_sent_at updates every time a field changes
Not sure why, but if a patient is marked `pledge_sent`, then `pledge_sent_at` and `pledge_sent_by` refreshes every single time there's an update. This is a bug in the `PatientsController#update` behavior -- see L63. Task here is to resolve that. Gold medal version is to move this to an activerecord callback (probably an `after_save` or whatever the 'while you're in this transaction, also' callback is?) if we can set the current user as part of it.

The ideal behavior is that information this is set when `pledge_sent` is, and nulled out when pledge sent is false or nil.

### PR DESCRIPTION
**Pull request by mdworken**
I rule and have completed some work on Case Manager that's ready for review!

Corrects logic to fix bug 1286.

### CODE REVIEW

File: app/controllers/patients_controller.rb [hunk 1]
```diff
@@ -119,7 +116,7 @@ def find_patient
                   :procedure_cost, :check_number, :date_of_check]
   ].freeze
 
-  OTHER_PARAMS = [:urgent_flag, :initial_call_date, :pledge_sent].freeze
+  OTHER_PARAMS = [:urgent_flag, :initial_call_date, :pledge_sent, :pledge_sent_by, :pledge_sent_at, :last_edited_by].freeze
```
Comment (colinxfleming) [inline app/controllers/patients_controller.rb]: hmm, I don't think you need the `pledge_sent_by`/`pledge_sent_at` here -- the `patient_params` deals explicitly with the payload sent to the controller for mass assignment (basically saying 'it's okay to change all of these fields at once, but only these fields'), and since I don't think we're setting them from the form, we can probably skip changing this set. Does that sound right to you?

File: app/models/patient.rb [hunk 2]
```diff
@@ -179,4 +182,14 @@ def clean_fields
   def initialize_fulfillment
     build_fulfillment(created_by_id: created_by_id).save
   end
+  
+  def update_sent_by_sent_at
+    if self.pledge_sent  && !self.pledge_sent_by
```
Comment (colinxfleming) [inline app/models/patient.rb]: This looks good! The following feedback is totally in the domain of style / rails-ifying things, so implement what makes sense to you:

* I don't believe we need the `self`s here -- the object is smart enough to know that it's referring to its own attributes, pleasantly enough. You can just write `pledge_sent` and it'll work like a charm, I do believe.
* One of the sickest features of rails is that it throws a ton of convenience methods which help make your programming look a lot more like readable english -- rather than do `!pledge_sent_by`, or `!pledge_sent` we can do `pledge_sent_by.blank?` to the same effect. [Docs here](http://api.rubyonrails.org/v4.2/classes/Object.html#method-i-blank-3F).

File: app/models/patient.rb [hunk 3]
```diff
@@ -38,7 +38,8 @@ class Patient
   embeds_many :notes
   belongs_to :pledge_generated_by, class_name: 'User', inverse_of: nil
   belongs_to :pledge_sent_by, class_name: 'User', inverse_of: nil
-
+  belongs_to :last_edited_by, class_name: 'User', inverse_of: nil
```
Comment (colinxfleming) [inline app/models/patient.rb]: I was initially on the fence about this but the more I think about it the more I like it. Good call on this! It's a good workaround to the problem of not having `current_user` in a model, it updates the stuff in its own db transaction, and we get a fringe benefit of having this in one convenient place if we need it. (We store something similar now but it's buried in the audit history, and wouldn't be available at the time the before_update callback fired.)

File: app/models/patient.rb [hunk 4]
```diff
@@ -26,7 +26,7 @@ class Patient
 
   before_validation :clean_fields
   before_save :save_identifier
-
+  before_update :update_sent_by_sent_at
```
Comment (colinxfleming) [inline app/models/patient.rb]: I think this name is close but doesn't quite capture the important thing, which is the pledge! Maybe `update_pledge_sent_by_sent_at`?

File: app/models/patient.rb [hunk 5]
```diff
@@ -160,6 +161,8 @@ def event_params
     }
   end
 
+
```
Comment (colinxfleming) [inline app/models/patient.rb]: Kill this whitespace please!

File: app/models/patient.rb [hunk 6]
```diff
@@ -179,4 +182,14 @@ def clean_fields
   def initialize_fulfillment
     build_fulfillment(created_by_id: created_by_id).save
   end
+  
+  def update_sent_by_sent_at
+    if self.pledge_sent  && !self.pledge_sent_by
+      self.pledge_sent_at = Time.zone.now
+      self.pledge_sent_by = last_edited_by
+    elsif !(self.pledge_sent)#       self.pledge_sent_at = nil
```
Comment (colinxfleming) [inline app/models/patient.rb]: I assume we can ditch this comment, yeah?

File: app/controllers/patients_controller.rb (hunk 1 above)
Comment (mdworken) [reply app/controllers/patients_controller.rb]: Yes, I think you're right!

File: app/models/patient.rb (hunk 4 above)
Comment (mdworken) [reply app/models/patient.rb]: Sounds good, will update it

File: app/models/patient.rb (hunk 6 above)
Comment (mdworken) [reply app/models/patient.rb]: Oops, yeah, that was in there from when I was debugging, haha

**Comment by mdworken** [comment]
Hi Colin!
Regarding the test failure, I would definitely expect it to have failed as the setting of pledge _sent_by and pledge_sent_at is no longer in the controller. I'll see if that change fixes it.

Re: model test, I haven't actually written model tests before, but I'll give it a crack late today after I merge in the other requested changes.

File: app/models/patient.rb (hunk 2 above)
Comment (mdworken) [reply app/models/patient.rb]: Upon testing, we don't need the self in the if and elsif statements, but we do when setting the variables. Something something scoping something?

**Comment by mdworken** [comment]
I sent slack messages, but the test that was previously passing seems to have had some issues. 

I suspect what was happening was:
1) the controller was updating the pledge_sent_at for the @patient, which led to the assert_kind_of Time to return true
2) the update failed validation for the database because it was missing clinic and fund pledge
3) the test that update works did NOT include a payload of pledge_sent, so no one noticed the database failing.

File: app/controllers/patients_controller.rb [hunk 7]
```diff
@@ -119,7 +116,7 @@ def find_patient
                   :procedure_cost, :check_number, :date_of_check]
   ].freeze
 
-  OTHER_PARAMS = [:urgent_flag, :initial_call_date, :pledge_sent].freeze
+  OTHER_PARAMS = [:urgent_flag, :initial_call_date, :pledge_sent, :last_edited_by].freeze
```
Comment (colinxfleming) [inline app/controllers/patients_controller.rb]: `last_edited_by` we draw from the `current_user` variable in the controller, not the payload params, so we can revert this change.

Comment (mdworken) [reply app/controllers/patients_controller.rb]: Good point!
